### DISCUSSION TITLE
asynchronous onBeforeRender causes timeout

### DISCUSSION THREAD

**Discussion by BanDroid** [vikejs/vike#971] [Discussion] [answered]
### Description

i am following [react-full](https://github.com/brillout/vite-plugin-ssr/tree/main/examples/react-full) example, altough it is working fine in dev, but in production, `onBeforeRender` causes warning and Timeout error.

```
Jun 24 02:32:11 PM  [vite-plugin-ssr][Warning] The onBeforeRender() hook defined by /src/pages/posts/@id.page.jsx > `export { onBeforeRender }` is taking more than 4 seconds
Jun 24 02:32:47 PM  Error: [vite-plugin-ssr][Error] Hook timeout: the onBeforeRender() hook defined by /src/pages/posts/@id.page.jsx > `export { onBeforeRender }` didn't finish after 40 seconds
Jun 24 02:32:47 PM      at Timeout._onTimeout (/opt/render/project/src/node_modules/vite-plugin-ssr/dist/cjs/utils/executeHook.js:39:50)
```

this made my `/posts/@id` route have 500 server internal error only in this route.

**Comment by brillout** [comment]
I don't understand the problem?

**Reply by BanDroid** [reply]
i am following the examples on making the title page dynamic using onBeforeRender, but it seems onBeforeRender loads a very long time in production.

i host it with render in free tier account. do you think it could cause the problem? because in free tier only have 512mb ram and 0.1 CPU

**Reply by brillout** [reply]
I don't think it's related to VPS.

**Reply by BanDroid** [reply]
try the [demo](https://vite-plugin-ssr-demo.onrender.com/) here, create a new post and see the post detail.

**Reply by brillout** [reply]
I really doubt it's related to VPS.

**Reply by BanDroid** [reply]
anything related to onBeforeRender function?

**Comment by BanDroid** [comment, marked as answer]
i found the problem, onBeforeRender causes infinite loop in production mode if fetching local data inside it, since onBeforeRender is running on server, i could directly query data from my databases in there, but i have created a new discussion about this to discuss new problem i found which is onBeforeRender showing the page very slow.

**Reply by brillout** [reply]
I'm not sure I see the problem. I'm inclined to think it's a user land thing (but feel free to let us know if you believe something about vite-plugin-ssr should be improved).
